File: app/lib/features/authentication/domain/extensions/liveness_session_entity_reference.dart
import 'package:app/features/authentication/domain/entities/liveness_session_entity.dart';
import 'package:app/features/users/domain/extensions/user_entity_reference.dart';
import 'package:cloud_firestore/cloud_firestore.dart';

extension LivenessSessionEntityReference on LivenessSessionEntity {
  static const String sessionsSubcollection = 'LivenessSessions';

  static DocumentReference<Map<String, dynamic>> firebaseSessionDocReference(
    FirebaseFirestore firestore,
    String uid,
    String sessionId,
  ) {
    return firebaseSessionsCollectionReference(firestore, uid).doc(sessionId);
  }

  static CollectionReference<Map<String, dynamic>>
      firebaseSessionsCollectionReference(
    FirebaseFirestore firestore,
    String uid,
  ) {
    return UserEntityReference.firebaseUidReference(firestore, uid)
        .collection(sessionsSubcollection);
  }

  static String cachedKey() => 'CACHED_LIVENESS_SESSION_INFO';

  static final List<String> sessionFields = [
    'userId',
    'provider',
    'providerSessionId',
    'status',
    'resultSummary',
    'rawPayloadRef',
    'createdAt',
    'completedAt',
  ];
}

File: backend/functions/src/features/authentication/domain/extensions/liveness_session_entity_reference.ts
import type {
  CollectionReference,
  DocumentData,
  DocumentReference,
  Firestore,
} from "firebase-admin/firestore";

import {UserEntityReference} from "./user_entity_reference.js";

/** Espelho de `LivenessSessionEntityReference` do app. */
export const LivenessSessionEntityReference = {
  sessionsSubcollection: "LivenessSessions",

  firebaseSessionDocReference(
    firestore: Firestore,
    uid: string,
    sessionId: string,
  ): DocumentReference<DocumentData> {
    return LivenessSessionEntityReference.firebaseSessionsCollectionReference(
      firestore,
      uid,
    ).doc(sessionId);
  },

  firebaseSessionsCollectionReference(
    firestore: Firestore,
    uid: string,
  ): CollectionReference<DocumentData> {
    return UserEntityReference.firebaseUidReference(firestore, uid).collection(
      LivenessSessionEntityReference.sessionsSubcollection,
    );
  },

  cachedKey(): string {
    return "CACHED_LIVENESS_SESSION_INFO";
  },

  sessionFields: [
    "userId",
    "provider",
    "providerSessionId",
    "status",
    "resultSummary",
    "rawPayloadRef",
    "createdAt",
    "completedAt",
  ] as const,
};
